src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Mime;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public static class HttpContentJsonExtensions
+    {
+        public static Task<object?> ReadFromJsonAsync(this HttpContent content, Type type, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore(content, type, options, cancellationToken);
+        }
+
+        public static Task<T> ReadFromJsonAsync<T>(this HttpContent content, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore<T>(content, options, cancellationToken);
+        }
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            byte[] contentBytes = await GetUtf8JsonBytesFromContentAsync(content, cancellationToken).ConfigureAwait(false);
+            return JsonSerializer.Deserialize(contentBytes, type, options);
+        }
+
+        private static async Task<T> ReadFromJsonAsyncCore<T>(HttpContent content, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            byte[] contentBytes = await GetUtf8JsonBytesFromContentAsync(content, cancellationToken).ConfigureAwait(false);
+            return JsonSerializer.Deserialize<T>(contentBytes, options);
+        }
+
+        private static async Task<byte[]> GetUtf8JsonBytesFromContentAsync(HttpContent content, CancellationToken cancellationToken)
+        {
+            string? mediaType = content.Headers.ContentType?.MediaType;
+
+            if (mediaType != JsonContent.JsonMediaType &&
+                mediaType != MediaTypeNames.Text.Plain)
+            {
+                throw new NotSupportedException("The provided ContentType is not supported; the supported types are 'application/json' and 'text/plain'.");
+            }
+
+            // Code taken from https://source.dot.net/#System.Net.Http/System/Net/Http/HttpContent.cs,047409be2a4d70a8
```

Comment:
Please try to move this code into the Common directory so we can avoid duplicating it.

Reply:
That code has semantics of determining the encoding out of the content. 
Do we want that to apply here as well?

Reply:
It looks like this code was removed from implementation, so original comment no longer applies.

Reply:
It is still there, I just moved it to a separate method called `GetEncoding`, see [here](https://github.com/dotnet/runtime/pull/33459/files#diff-75e20d2a014e7c134181d017ed9f92d6R66-R93). Sorry for not making that clear.

Compared with the logic used to determine the encoding in `HttpContent.ReadAsStringAsync` (see [here](https://source.dot.net/#System.Net.Http/System/Net/Http/HttpContent.cs,047409be2a4d70a8)), which is very similar to `GetEncoding` method in this class.

Do we want to apply the same rules here? More precisely, do we want to use the BOM to determine the encoding when is not present in the response? Is it safe? cc @GrabYourPitchforks 

@rynowak has any BOM usage shown up in users requests/responses?

Reply:
I don't see a problem with the logic in `GetEncoding`. There's a few issues at play.

The default for `application/json` is UTF8 (if not otherwise specified)
It's against spec to put a BOM in `application/json` content

Therefore trying to sniff an encoding is a case that's only hit when you have a non-compliant server. 

> has any BOM usage shown up in users requests/responses?

There's a really easy way to hit a case like what I described above:

- Create a file on disk saved with a UTF8 BOM, call it `pokemon.json`
- Set up a webserver to serve that file

Your webserver will likely assign `application/json` to the response based on the file extension (ASP.NET Core and IIS both do). Your webserver will likely pass through the entire file contents including the BOM.

Reply:
@rynowak the [latest json RFC](https://tools.ietf.org/html/rfc8259) (there have been a few updates) states:

> JSON text exchanged between systems that are not part of a closed ecosystem MUST be encoded using UTF-8 [RFC3629].
>
> Previous specifications of JSON have not required the use of UTF-8 when transmitting JSON text.  However, the vast majority of JSON-based software implementations have chosen to use the UTF-8 encoding, to the extent that it is the only encoding that achieves interoperability.
>
> **Implementations MUST NOT add a byte order mark (U+FEFF) to the beginning of a networked-transmitted JSON text.  In the interests of interoperability, implementations that parse JSON texts MAY ignore the presence of a byte order mark rather than treating it as an error.**

Reply:
@rynowak I didn't interpret RFC 4627, Sec. 3, to mandate encoding sniffing. It simply says "SHALL be encoded in Unicode." It gives a technique for implementers to sniff the encoding if they desire, but there's no text here precluding an implementation from using a sidecar mechanism (such as the _Content-Type_ header) to determine the proper encoding to use. I think ASP.NET's behavior is fine. :)

Re: byte-order marks, I'm impartial as to whether we support them. In my experience I've not seen web sites specify _both_ a charset _and_ a BOM in the response. Even in the "reading files from disk" scenario, most web servers realize that they're acting as a pass-through and shouldn't make judgment calls about what the proper charset of the content should be.

If we do support byte-order marks, we shouldn't sniff the BOM itself. Instead, we should continue to rely on the incoming charset, stripping out the BOM _only if_: (a) it matches the selected encoding's BOM; _and_ (b) it occurs at the beginning of the payload.

But honestly I wouldn't add this behavior unless somebody's asking for it. It's just going to complicate matters.